# Introduction to Python

Forked from [Lecture 1](https://github.com/jrjohansson/scientific-python-lectures/blob/master/Lecture-1-Introduction-to-Python-Programming.ipynb) of [Scientific Python Lectures](http://github.com/jrjohansson/scientific-python-lectures) by [J.R. Johansson](http://jrjohansson.github.io/)

## Python Program Files

* Python code is usually stored in text files with the file ending in "`.py`":
  ```
  myprogram.py
  ```
* Every line in a Python program file is assumed to be a Python statement, or part thereof. 
    * The only exception is comment lines, which start with the character `#` (optionally preceded by an arbitrary number of white-space characters, i.e., tabs or spaces). Comment lines are usually ignored by the Python interpreter.
    ```
    # this is a comment
    ```
* To run our Python program from the command line we use:
  ```
  $ python myprogram.py
  ```
* On UNIX systems it is common to define the path to the interpreter on the first line of the program (note that this is a comment line as far as the Python interpreter is concerned):
  ```
  #!/usr/bin/env python
  ```

  If we do, and if we additionally set the file script to be executable, we can run the program like this:
  ```
  $ myprogram.py
  ```

In [1]:
print("hello world")

hello world


### Example:

In [3]:
!ls /Users/mahmoud/Desktop/

MLNN Videos                     mlnn
Week 1 - Jupyter Notebook.ipynb myprogram.py
cpython                         untitled folder


In [2]:
!ls ../scripts/hello-world*.py

../scripts/hello-world.py


In [5]:
!cat /Users/mahmoud/Desktop/myprogram.py

print("hello world")
print(2+3)

In [6]:
!python /Users/mahmoud/Desktop/myprogram.py

hello world
5


## Jupyter Notebooks

This file - a Jupyter (IPython) notebook - does not follow the standard pattern with Python code in a text file. Instead, an IPython notebook is stored as a file in the [JSON](http://en.wikipedia.org/wiki/JSON) format. The advantage is that we can mix formatted text, Python code and code output. It requires the IPython notebook server to run it though, and therefore isn't a stand-alone Python program as described above. Other than that, there is no difference between the Python code that goes into a program file or an IPython notebook.

## Modules

Most of the functionality in Python is provided by *modules*. The Python Standard Library is a large collection of modules that provides *cross-platform* implementations of common facilities such as access to the operating system, file I/O, string management, network communication, and much more.

### References

 * The Python Language Reference: https://docs.python.org/3/reference/index.html
 * The Python Standard Library: https://docs.python.org/3/library/

To use a module in a Python program it first has to be imported. A module can be imported using the `import` statement. For example, to import the module `math`, which contains many standard mathematical functions, we can do:

In [7]:
import math

This includes the whole module and makes it available for use later in the program. For example, we can do:

In [9]:
math.cos(0)

1.0

In [10]:
import math
x = math.cos(2 * math.pi)
print(x)

1.0


Alternatively, we can chose to import all symbols (functions and variables) in a module to the current namespace (so that we don't need to use the prefix "`math.`" every time we use something from the `math` module:

In [6]:
from math import *
x = cos(2 * pi)
print(x)

1.0


This pattern can be very convenient, but in large programs that include many modules it is often a good idea to keep the symbols from each module in their own namespaces, by using the `import math` pattern. This would elminate potentially confusing problems with name space collisions.

As a third alternative, we can chose to import only a few selected symbols from a module by explicitly listing which ones we want to import instead of using the wildcard character `*`:

In [17]:
from math import cos, pi, sin
x = cos(2 * pi)
sin(0)
print(x)
math.

1.0


### Looking at what a module contains, and its documentation

Once a module is imported, we can list the symbols it provides using the `dir` function:

In [8]:
import math
print(dir(math))

['__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'pi', 'pow', 'radians', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'trunc']


And using the function `help` we can get a description of each function (almost .. not all functions have docstrings, as they are technically called, but the vast majority of functions are documented this way). 

In [20]:
help(math.pi)

Help on float object:

class float(object)
 |  float(x) -> floating point number
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(...)
 |      float.__format__(format_spec) -> string
 |      
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getformat__(...) from builtins.type
 |      float.__getformat__(typestr) -> string
 |      
 |      You probably don't want to use thi

In [24]:
math.log(10)/math.log(2)

3.3219280948873626

In [25]:
2**3.3219280948873626

10.000000000000002

In [26]:
math.log(10, 2.5)

2.51294159473206

In [28]:
help(math)

Help on module math:

NAME
    math

MODULE REFERENCE
    https://docs.python.org/3.5/library/math.html
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This module is always available.  It provides access to the
    mathematical functions defined by the C standard.

FUNCTIONS
    acos(...)
        acos(x)
        
        Return the arc cosine (measured in radians) of x.
    
    acosh(...)
        acosh(x)
        
        Return the inverse hyperbolic cosine of x.
    
    asin(...)
        asin(x)
        
        Return the arc sine (measured in radians) of x.
    
    asinh(...)
        asinh(x)
        
        Return the inverse hyperbolic sine of x.
    
    atan(...)
        atan(x)
     

We can also use the `help` function directly on modules: Try

    help(math) 

Some very useful modules form the Python standard library are `os`, `sys`, `math`, `shutil`, `re`, `subprocess`, `multiprocessing`, `threading`. 

A complete lists of standard modules for Python 3 are available at http://docs.python.org/3/library/.

For example, this is the `os` module in the standard library.

In [36]:
import os
desktop = os.listdir('/Users/mahmoud/Desktop/')
desktop[4]

'cpython'

## Variables and types

### Symbol names 

Variable names in Python can contain alphanumerical characters `a-z`, `A-Z`, `0-9` and some special characters such as `_`. Normal variable names must start with a letter. 

By convention, variable names start with a lower-case letter, and Class names start with a capital letter. 

In addition, there are a number of Python keywords that cannot be used as variable names. These keywords are:

    and, as, assert, break, class, continue, def, del, elif, else, except, 
    exec, finally, for, from, global, if, import, in, is, lambda, not, or,
    pass, print, raise, return, try, while, with, yield


In [43]:
poop_emoji = "💩"
print("hello" + poop_emoji)


hello💩


### Assignment

The assignment operator in Python is `=`. Python is a dynamically typed language, so we do not need to specify the type of a variable when we create one.

Assigning a value to a new variable creates the variable:

In [46]:
# variable assignments
x = 1.0
my_variable = 12.2

Although not explicitly specified, a variable does have a type associated with it. The type is derived from the value that was assigned to it.

In [14]:
type(x)

float

If we assign a new value to a variable, its type can change.

In [48]:
x = 1

In [49]:
type(x)

int

If we try to use a variable that has not yet been defined we get an `NameError`:

In [50]:
import traceback 

try:
    print(y)
except NameError as e:
    print(traceback.format_exc())

Traceback (most recent call last):
  File "<ipython-input-50-0488135cf974>", line 4, in <module>
    print(y)
NameError: name 'y' is not defined



In [51]:
y

NameError: name 'y' is not defined

### Fundamental types

In [18]:
# integers
x = 1
type(x)

int

In [19]:
# float
x = 1.0
type(x)

float

In [20]:
# boolean
b1 = True
b2 = False

type(b1)

bool

In [21]:
# complex numbers: note the use of `j` to specify the imaginary part
x = 1.0 - 1.0j
type(x)

complex

In [22]:
print(x)

(1-1j)


In [23]:
print(x.real, x.imag)

1.0 -1.0


### Type utility functions

In [52]:
x = 1.0

# check if the variable x is a float
type(x) is float

True

In [53]:
# check if the variable x is an int
type(x) is int

False

We can also use the `isinstance` method for testing types of variables:

In [54]:
isinstance(x, float)

True

### Type casting

In [55]:
x = 1.5

print(x, type(x))

1.5 <class 'float'>


In [28]:
x = int(x)

print(x, type(x))

1 <class 'int'>


In [56]:
z = complex(x)
z = 1 + 2j
print(z, type(z))

(1+2j) <class 'complex'>


In [61]:
z.real

1.0

In [62]:
z.imag

2.0

In [57]:
import traceback 

try:
    x = float(z)
except TypeError as e:
    print(traceback.format_exc())

Traceback (most recent call last):
  File "<ipython-input-57-68b971d0047b>", line 4, in <module>
    x = float(z)
TypeError: can't convert complex to float



## Operators and comparisons

Most operators and comparisons in Python work as one would expect:

* Arithmetic operators `+`, `-`, `*`, `/`, `//` (integer division), '**' power


In [64]:
7//3

2

In [31]:
1 + 2, 1 - 2, 1 * 2, 1 / 2

(3, -1, 2, 0.5)

In [32]:
1.0 + 2.0, 1.0 - 2.0, 1.0 * 2.0, 1.0 / 2.0

(3.0, -1.0, 2.0, 0.5)

In [33]:
# Integer division of float numbers
3.0 // 2.0

1.0

In [34]:
# Note! The power operators in python isn't ^, but **
2 ** 2

4

Note: The `/` operator always performs a floating point division in Python 3.x.
This is not true in Python 2.x, where the result of `/` is always an integer if the operands are integers.
to be more specific, `1/2 = 0.5` (`float`) in Python 3.x, and `1/2 = 0` (`int`) in Python 2.x (but `1.0/2 = 0.5` in Python 2.x).

The boolean operators are spelled out as the words `and`, `not`, `or`. 

In [65]:
True and False

False

In [66]:
not False

True

In [67]:
True or False

True

Comparison operators `>`, `<`, `>=` (greater or equal), `<=` (less or equal), `==` equality, `is` identical.

In [38]:
2 > 1, 2 < 1

(True, False)

In [39]:
2 > 2, 2 < 2

(False, False)

In [40]:
2 >= 2, 2 <= 2

(True, True)

In [41]:
# equality
[1,2] == [1,2]

True

In [70]:
# objects identical?
list1 = list2 = [1,2]
list3 = [1,2]
# list1 is list2
list3 == list2

True

**Exercise:**

Mindy has $5.25 in her pocket. Apples cost 29 cents each. Calculate how many apples mindy can buy and how much change she will have left. Money should be represented in variables of type float and apples should be represented in variables of type integer.

Answer:

- mindy_money = 5.25
- apple_cost = .29
- num_apples = ?
- change = ?

In [77]:
mindy_money = 5.25
num_apples = mindy_money//.29
num_apples

18.0

In [78]:
change = mindy_money - (num_apples*.29)
change

0.03000000000000025

## Compound types: Strings, List and dictionaries

### Strings

Strings are the variable type that is used for storing text messages. 

In [79]:
s = "Hello world"
type(s)

str

In [44]:
# length of the string: the number of characters
len(s)

11

In [81]:
# replace a substring in a string with somethign else
s2 = s.replace("world", "test")
print(s2)

Hello test


We can index a character in a string using `[]`:

In [83]:
s[1]

'e'

**Heads up MATLAB and R users:** Indexing start at 0!

We can extract a part of a string using the syntax `[start:stop]`, which extracts characters between index `start` and `stop` -1 (the character at index `stop` is not included):

In [47]:
s[0:5]

'Hello'

In [48]:
s[4:5]

'o'

If we omit either (or both) of `start` or `stop` from `[start:stop]`, the default is the beginning and the end of the string, respectively:

In [49]:
s[:5]

'Hello'

In [89]:
s[-2:]

'ld'

In [51]:
s[:]

'Hello world'

We can also define the step size using the syntax `[start:end:step]` (the default value for `step` is 1, as we saw above):

In [87]:
s[:6:1]

'Hello '

In [85]:
s[0:6:2]

'Hlo'

This technique is called *slicing*. Read more about the syntax here: https://docs.python.org/3.5/library/functions.html#slice

Python has a very rich set of functions for text processing. See for example https://docs.python.org/3.5/library/string.html for more information.

#### String formatting examples

In [54]:
print("str1", "str2", "str3")  # The print statement concatenates strings with a space

str1 str2 str3


In [55]:
print("str1", 1.0, False, -1j)  # The print statements converts all arguments to strings

str1 1.0 False (-0-1j)


In [56]:
print("str1" + "str2" + "str3") # strings added with + are concatenated without space

str1str2str3


In [57]:
print("value = %f" % 1.0)       # we can use C-style string formatting

value = 1.000000


In [58]:
# this formatting creates a string
s2 = "value1 = %.2f. value2 = %d" % (3.1415, 1.5)

print(s2)

value1 = 3.14. value2 = 1


In [91]:
# alternative, more intuitive way of formatting a string 
s3 = 'value1 = {0}, value2 = {1}'.format(3.1415, 1.5)

print(s3)

value1 = 3.1415, value2 = 1.5


**Exercise:**

Paste in the code from your previous exercise and output the result as a story (round monetary values to 2 decimal places). The ouptut should look like this:

"Mindy had \$5.25 in her pocket. Apples at her nearby store cost 29 cents. With her \$5.25, mindy can buy 18 apples and will have 10 cents left over."

### List

Lists are very similar to strings, except that each element can be of any type.

The syntax for creating lists in Python is `[...]`:

In [60]:
l = [1,2,3,4]

print(type(l))
print(l)

<class 'list'>
[1, 2, 3, 4]


We can use the same slicing techniques to manipulate lists as we could use on strings:

In [61]:
print(l)

print(l[1:3])

print(l[::2])

[1, 2, 3, 4]
[2, 3]
[1, 3]


**Heads up MATLAB and R users:** Indexing starts at 0!

In [62]:
l[0]

1

Elements in a list do not all have to be of the same type:

In [63]:
l = [1, 'a', 1.0, 1-1j]

print(l)

[1, 'a', 1.0, (1-1j)]


Python lists can be heterogeneous and arbitrarily nested:

In [64]:
nested_list = [1, [2, [3, [4, [5]]]]]

nested_list

[1, [2, [3, [4, [5]]]]]

Lists play a very important role in Python. For example they are used in loops and other flow control structures (discussed below). There are a number of convenient functions for generating lists of various types, for example the `range` function:

In [65]:
start = 10
stop = 30
step = 2

range(start, stop, step)

range(10, 30, 2)

In [66]:
# in python 3 range generates an iterator, which can be converted to a list using 'list(...)'.
# It has no effect in python 2
list(range(start, stop, step))

[10, 12, 14, 16, 18, 20, 22, 24, 26, 28]

In [67]:
list(range(-10, 10))

[-10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [68]:
s

'Hello world'

In [69]:
# convert a string to a list by type casting:
s2 = list(s)

s2

['H', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd']

In [70]:
# sorting lists (by creating a new variable)

s3 = sorted(s2)

print(s2)
print(s3)

['H', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd']
[' ', 'H', 'd', 'e', 'l', 'l', 'l', 'o', 'o', 'r', 'w']


In [71]:
# sorting lists in place
s2.sort()

print(s2)

[' ', 'H', 'd', 'e', 'l', 'l', 'l', 'o', 'o', 'r', 'w']


#### Adding, inserting, modifying, and removing elements from lists

In [72]:
# create a new empty list
l = []

# add an elements using `append`
l.append("A")
l.append("d")
l.append("d")

print(l)

['A', 'd', 'd']


We can modify lists by assigning new values to elements in the list. In technical jargon, lists are *mutable*.

In [73]:
l[1] = "p"
l[2] = "p"

print(l)

['A', 'p', 'p']


In [74]:
l[1:3] = ["d", "d"]

print(l)

['A', 'd', 'd']


Insert an element at an specific index using `insert`

In [75]:
l.insert(0, "i")
l.insert(1, "n")
l.insert(2, "s")
l.insert(3, "e")
l.insert(4, "r")
l.insert(5, "t")

print(l)

['i', 'n', 's', 'e', 'r', 't', 'A', 'd', 'd']


Remove first element with specific value using 'remove'

In [76]:
l.remove("A")

print(l)

['i', 'n', 's', 'e', 'r', 't', 'd', 'd']


Remove an element at a specific location using `del`:

In [77]:
del l[7]
del l[6]

print(l)

['i', 'n', 's', 'e', 'r', 't']


See `help(list)` for more details, or read the online documentation 

### Tuples

Tuples are like lists, except that they cannot be modified once created, that is they are *immutable*. 

In Python, tuples are created using the syntax `(..., ..., ...)`, or even `..., ...`:

In [78]:
point = (10, 20)

print(point, type(point))

(10, 20) <class 'tuple'>


In [79]:
point = 10, 20

print(point, type(point))

(10, 20) <class 'tuple'>


We can unpack a tuple by assigning it to a comma-separated list of variables:

In [80]:
x, y = point

print("x =", x)
print("y =", y)

x = 10
y = 20


If we try to assign a new value to an element in a tuple we get an error:

In [81]:
try:
    point[0] = 20
except TypeError as e:
    print(traceback.format_exc())

Traceback (most recent call last):
  File "<ipython-input-81-c017eac57cd7>", line 2, in <module>
    point[0] = 20
TypeError: 'tuple' object does not support item assignment



### Dictionaries

Dictionaries are also like lists, except that each element is a key-value pair. The syntax for dictionaries is `{key1 : value1, ...}`:

In [82]:
params = {"parameter1" : 1.0,
          "parameter2" : 2.0,
          "parameter3" : 3.0,}

print(type(params))
print(params)

<class 'dict'>
{'parameter1': 1.0, 'parameter3': 3.0, 'parameter2': 2.0}


In [83]:
print("parameter1 = " + str(params["parameter1"]))
print("parameter2 = " + str(params["parameter2"]))
print("parameter3 = " + str(params["parameter3"]))

parameter1 = 1.0
parameter2 = 2.0
parameter3 = 3.0


In [84]:
params["parameter1"] = "A"
params["parameter2"] = "B"

# add a new entry
params["parameter4"] = "D"

print("parameter1 = " + str(params["parameter1"]))
print("parameter2 = " + str(params["parameter2"]))
print("parameter3 = " + str(params["parameter3"]))
print("parameter4 = " + str(params["parameter4"]))

parameter1 = A
parameter2 = B
parameter3 = 3.0
parameter4 = D


**Exercise:**

Mindy doesn't want 18 apples, that's too many for someone who lives by themselves. We're now going to represent mindy's world using our new data types.

Make a list containing the fruits that mindy desires. She likes apples, strawberries, pinapples, and papayas.

Make a tuple containing the fruits that the store has. This is immutable because the store doesn't change their inventory. The local store has apples, strawberries, pinapples, pears, bananas, and oranges.

Make a dictonary showing the price of each fruit at the store. Apples are 29 cents, bananas are 5 cents, oranges are 20 cents, strawberries are 30 cents and pinapples are $1.50.

## Control Flow

### Conditional statements: if, elif, else

The Python syntax for conditional execution of code uses the keywords `if`, `elif` (else if), `else`:

In [85]:
statement1 = False
statement2 = False

if statement1:
    print("statement1 is True")
    
elif statement2:
    print("statement2 is True")
    
else:
    print("statement1 and statement2 are False")

statement1 and statement2 are False


For the first time, here we encounted a peculiar and unusual aspect of the Python programming language: Program blocks are defined by their indentation level. 

Compare to the equivalent C code:

    if (statement1)
    {
        printf("statement1 is True\n");
    }
    else if (statement2)
    {
        printf("statement2 is True\n");
    }
    else
    {
        printf("statement1 and statement2 are False\n");
    }

In C blocks are defined by the enclosing curly brakets `{` and `}`. And the level of indentation (white space before the code statements) does not matter (completely optional). 

But in Python, the extent of a code block is defined by the indentation level (usually a tab or say four white spaces). This means that we have to be careful to indent our code correctly, or else we will get syntax errors. 

#### Examples:

In [86]:
statement1 = statement2 = True

if statement1:
    if statement2:
        print("both statement1 and statement2 are True")

both statement1 and statement2 are True


In [87]:
# # Bad indentation!
# if statement1:
#     if statement2:
#     print("both statement1 and statement2 are True")  # this line is not properly indented

In [88]:
statement1 = False 

if statement1:
    print("printed if statement1 is True")
    
    print("still inside the if block")

In [89]:
if statement1:
    print("printed if statement1 is True")
    
print("now outside the if block")

now outside the if block


## Loops

In Python, loops can be programmed in a number of different ways. The most common is the `for` loop, which is used together with iterable objects, such as lists. The basic syntax is:

### **`for` loops**:

In [90]:
for x in [1,2,3]:
    print(x)

1
2
3


The `for` loop iterates over the elements of the supplied list, and executes the containing block once for each element. Any kind of list can be used in the `for` loop. For example:

In [91]:
for x in range(4): # by default range start at 0
    print(x)

0
1
2
3


Note: `range(4)` does not include 4 !

In [92]:
for x in range(-3,3):
    print(x)

-3
-2
-1
0
1
2


In [93]:
for word in ["scientific", "computing", "with", "python"]:
    print(word)

scientific
computing
with
python


To iterate over key-value pairs of a dictionary:

In [94]:
for key, value in params.items():
    print(key + " = " + str(value))

parameter1 = A
parameter4 = D
parameter3 = 3.0
parameter2 = B


Sometimes it is useful to have access to the indices of the values when iterating over a list. We can use the `enumerate` function for this:

In [95]:
for idx, x in enumerate(range(-3,3)):
    print(idx, x)

0 -3
1 -2
2 -1
3 0
4 1
5 2


### List comprehensions: Creating lists using `for` loops:

A convenient and compact way to initialize lists:

In [96]:
l1 = [x**2 for x in range(0,5)]

print(l1)

[0, 1, 4, 9, 16]


### `while` loops:

In [97]:
i = 0

while i < 5:
    print(i)
    
    i = i + 1
    
print("done")

0
1
2
3
4
done


Note that the `print("done")` statement is not part of the `while` loop body because of the difference in indentation.

**Exercise:**

Loop through all of the fruits that mindy wants and check if the store has them. For each fruit that she wants print 

"Mindy, the store has apples and they cost $.29"

or

"Mindy, the store does not have papayas"

## Functions

A function in Python is defined using the keyword `def`, followed by a function name, a signature within parentheses `()`, and a colon `:`. The following code, with one additional level of indentation, is the function body.

In [95]:
def func0(s):   
    print(s + "class")

In [96]:
func0("hello ")

hello class


Optionally, but highly recommended, we can define a so called "docstring", which is a description of the functions purpose and behaivor. The docstring should follow directly after the function definition, before the code in the function body.

In [97]:
def func1(s):
    """
    Print a string 's' and tell how many characters it has    
    """
    
    print(s + " has " + str(len(s)) + " characters")

In [100]:
help(func1)

Help on function func1 in module __main__:

func1(s)
    Print a string 's' and tell how many characters it has



In [106]:
result = func1("testing your logic")

testing your logic has 18 characters


In [107]:
result

Functions that returns a value use the `return` keyword:

In [102]:
def square(x):
    """
    Return the square of x.
    """
    return x ** 2

In [ ]:
y = f(x)
y = x**2

In [104]:
number = square(4)

In [105]:
number

16

We can return multiple values from a function using tuples (see above):

In [112]:
def powers(x):
    """
    Return a few powers of x.
    """
    return x ** 2, x ** 3, x ** 4

In [113]:
powers(3)

(9, 27, 81)

In [115]:
x2, x3, x4 = powers(3)

print(x2)

9


### Default argument and keyword arguments

In a definition of a function, we can give default values to the arguments the function takes:

In [108]:
def myfunc(x, p=2, debug=False):
    if debug:
        print("evaluating myfunc for x = " + str(x) + " using exponent p = " + str(p))
    return x**p

If we don't provide a value of the `debug` argument when calling the the function `myfunc` it defaults to the value provided in the function definition:

In [109]:
myfunc(5)

25

In [110]:
myfunc(5, debug=True)

evaluating myfunc for x = 5 using exponent p = 2


25

If we explicitly list the name of the arguments in the function calls, they do not need to come in the same order as in the function definition. This is called *keyword* arguments, and is often very useful in functions that takes a lot of optional arguments.

In [111]:
myfunc(p=3, debug=True, x=7)

evaluating myfunc for x = 7 using exponent p = 3


343

### Unnamed functions (lambda function)

In Python we can also create unnamed functions, using the `lambda` keyword:

In [112]:
f1 = lambda x: x**2
    
# is equivalent to 

def f2(x):
    return x**2

In [113]:
f1(2), f2(2)

(4, 4)

This technique is useful for example when we want to pass a simple function as an argument to another function, like this:

In [114]:
# map is a built-in python function
map(lambda x: x**2, range(-3,4))

In [115]:
# in python 3 we can use `list(...)` to convert the iterator to an explicit list
list(map(lambda x: x**2, range(-3,4)))

[9, 4, 1, 0, 1, 4, 9]

**Exercise:**

Mindy is great, but we want code that can tell anyone what fruits the store has. To do this we will generalize our code for Mindy using a function.

Write a function that takes the following parameters
- full_name (string)
- fruits_you_want (list)
- fruits_the_store_has (tuple)
- prices (dict)

and prints to the terminal a sentence per fruit that you want just like the last exercise. For example, if 

```
name = 'Al'
list_of_fruits_you_want = ['apple', 'banana']
tuple_of_fruits_the_store_has = ('apple', 'banana', 'orange', 'strawberries', 'pineapple')
prices = {
    'apple' : .29
    'banana':  .05
    'orange': .20
    'strawberries': .30
    'pinapple': 1.50
}
```

The function should print.

"Al, the store has apples and they cost \$.29"  
"Al, the store has bananas and they cost \$.05"

## Classes

Classes are the key features of object-oriented programming. A class is a structure for representing an object and the operations that can be performed on the object. 

In Python a class can contain *attributes* (variables) and *methods* (functions).

A class is defined almost like a function, but using the `class` keyword, and the class definition usually contains a number of class method definitions (a function in a class).

* Each class method should have an argument `self` as its first argument. This object is a self-reference.

* Some class method names have special meaning, for example:

    * `__init__`: The name of the method that is invoked when the object is first created.
    * `__str__` : A method that is invoked when a simple string representation of the class is needed, as for example when printed.
    * There are many more, see http://docs.python.org/2/reference/datamodel.html#special-method-names

In [116]:
class Point:
    """
    Simple class for representing a point in a Cartesian coordinate system.
    """
    
    def __init__(self, x, y):
        """
        Create a new Point at x, y.
        """
        self.x = x
        self.y = y
        
    def translate(self, dx, dy):
        """
        Translate the point by dx and dy in the x and y direction.
        """
        self.x += dx
        self.y += dy
        
    def __str__(self):
        return("Point at [%f, %f]" % (self.x, self.y))

To create a new instance of a class:

In [117]:
p1 = Point(0, 0) # this will invoke the __init__ method in the Point class

print(p1)         # this will invoke the __str__ method

Point at [0.000000, 0.000000]


To invoke a class method in the class instance `p`:

In [118]:
p2 = Point(1, 1)

p1.translate(0.25, 1.5)

print(p1)
print(p2)

Point at [0.250000, 1.500000]
Point at [1.000000, 1.000000]


Note that calling class methods can modifiy the state of that particular class instance, but does not effect other class instances or any global variables.

That is one of the nice things about object-oriented design: code such as functions and related variables are grouped in separate and independent entities. 

## Modules

One of the most important concepts in good programming is to reuse code and avoid repetitions.

The idea is to write functions and classes with a well-defined purpose and scope, and reuse these instead of repeating similar code in different part of a program (modular programming). The result is usually that readability and maintainability of a program is greatly improved. What this means in practice is that our programs have fewer bugs, are easier to extend and debug/troubleshoot. 

Python supports modular programming at different levels. Functions and classes are examples of tools for low-level modular programming. Python modules are a higher-level modular programming construct, where we can collect related variables, functions and classes in a module. A python module is defined in a python file (with file-ending `.py`), and it can be made accessible to other Python modules and programs using the `import` statement. 

Consider the following example: the file `mymodule.py` contains simple example implementations of a variable, function and a class:

In [119]:
%%file mymodule.py
"""
Example of a python module. Contains a variable called my_variable,
a function called my_function, and a class called MyClass.
"""

my_variable = 0

def my_function():
    """
    Example function
    """
    return my_variable
    
class MyClass:
    """
    Example class.
    """

    def __init__(self):
        self.variable = my_variable
        
    def set_variable(self, new_value):
        """
        Set self.variable to a new value
        """
        self.variable = new_value
        
    def get_variable(self):
        return self.variable

Overwriting mymodule.py


We can import the module `mymodule` into our Python program using `import`:

In [120]:
import mymodule

Use `help(module)` to get a summary of what the module provides:

In [121]:
help(mymodule)

Help on module mymodule:

NAME
    mymodule

DESCRIPTION
    Example of a python module. Contains a variable called my_variable,
    a function called my_function, and a class called MyClass.

CLASSES
    builtins.object
        MyClass
    
    class MyClass(builtins.object)
     |  Example class.
     |  
     |  Methods defined here:
     |  
     |  __init__(self)
     |      Initialize self.  See help(type(self)) for accurate signature.
     |  
     |  get_variable(self)
     |  
     |  set_variable(self, new_value)
     |      Set self.variable to a new value
     |  
     |  ----------------------------------------------------------------------
     |  Data descriptors defined here:
     |  
     |  __dict__
     |      dictionary for instance variables (if defined)
     |  
     |  __weakref__
     |      list of weak references to the object (if defined)

FUNCTIONS
    my_function()
        Example function

DATA
    my_variable = 0

FILE
    /Users/johria/Development/DAT-DC

In [122]:
mymodule.my_variable

0

In [123]:
mymodule.my_function() 

0

In [124]:
my_class = mymodule.MyClass() 
my_class.set_variable(10)
my_class.get_variable()

10

If we make changes to the code in `mymodule.py`, we need to reload it using `reload`:

In [125]:
import importlib
importlib.reload(mymodule) # Python 3 only
# For Python 2 use reload(mymodule)

<module 'mymodule' from '/Users/johria/Development/DAT-DC-12/notebooks/mymodule.py'>

## Exceptions

In Python errors are managed with a special language construct called "Exceptions". When errors occur exceptions can be raised, which interrupts the normal program flow and fallback to somewhere else in the code where the closest try-except statement is defined.

To generate an exception we can use the `raise` statement, which takes an argument that must be an instance of the class `BaseException` or a class derived from it. 

In [126]:
try:
    raise Exception("description of the error")
except Exception as e:
    print(traceback.format_exc())

Traceback (most recent call last):
  File "<ipython-input-126-72b44174a06f>", line 2, in <module>
    raise Exception("description of the error")
Exception: description of the error



A typical use of exceptions is to abort functions when some error condition occurs, for example:

    def my_function(arguments):
    
        if not verify(arguments):
            raise Exception("Invalid arguments")
        
        # rest of the code goes here

To gracefully catch errors that are generated by functions and class methods, or by the Python interpreter itself, use the `try` and  `except` statements:

    try:
        # normal code goes here
    except:
        # code for error handling goes here
        # this code is not executed unless the code
        # above generated an error

For example:

In [127]:
try:
    print("test")
    # generate an error: the variable test is not defined
    print(test)
except Exception:
    print("Caught an exception")

test
Caught an exception


To get information about the error, we can access the `Exception` class instance that describes the exception by using for example:

    except Exception as e:

In [128]:
try:
    print("test")
    # generate an error: the variable test is not defined
    print(test)
except Exception as e:
    print("Caught an exception:", e)

test
Caught an exception: name 'test' is not defined


**Excercise:**

Make two classes with the following variables and methods

**Store**

variables
- inventory (dict)

methods
- `show_inventory()`
 * nicely displays the store's inventory
- `message_customer(customer)`
 * shows the customer if the store has the fruits they want and how much each fruit costs (this is code from previous exercises, except it will use the customer's "Formal Greeting" instead of their first name.)  

**Customer** 

variables
- `first_name` (string)
- `last_name` (string)
- `is_male` (boolean)
- `money` (float)
- `fruit` (dict)
- `preferred_fruit` (list)

methods
- `formal_greeting()`
    * (Mr. Al Johri, Ms. Mindy Smith)
- `buy_fruit(store, fruit_name, fruit_amt)`
    * inputs are a store, the name of a fruit, and the amount of that fruit
    * checks to see if the store has the fruit - returns an error if it does not
    * checks to see if the customer can afford the amount of fruit they intend to buy - returns an error if not
    * "purchases" the fruit by adding it to the customers fruit dict and removes the correct amount of money from their money variable
    

**Exercise:**

Instantiate a list of the following customers.

* Mindy Smith - \$5.25, likes apples and oranges
* Al Johri - \$20.19, likes papaya, strawberries, pinapple, and apples  
* Hillary Clinton - \$15, likes strawberries and oranges  
* Oliver Twist - \$.05, likes apples  
* Donald Trump - \$4000, only likes durian

Create a store called Whole Foods with the following inventory
   * 'apple' : \$.29
   * 'banana':  \$.05
   * 'orange': \$.20
   * 'strawberries': \$.30
   * 'pinapple': \$1.50,
   * 'grapes': \$.22,
   * 'durian': \$5000

Write code to do the following

1. Print the store's inventory
2. For each customer, print the store's message to them
3. Have each customer purchase 1 of each fruit in their list of preferred fruits. 

(Make sure you have error handling so the program doesn't halt if the store doesn't have the fruit the customer wants or if the customer doesn not have enough money to buy the fruit.)
    


**Bonus Exercise**

Organize the code! Make a module called `fruits` (in a file fruits.py) that contains the class definitions.
Make a separate cell (in a file main.py) which imports the fruits module, instantiates the store and the list of customers, and runs the code in the previous exercise.


## Further reading

* http://www.python.org - The official web page of the Python programming language.
* https://docs.python.org/3/tutorial/ - The Official Python Tutorial
* http://www.python.org/dev/peps/pep-0008 - Style guide for Python programming. Highly recommended. 
* http://www.scipy-lectures.org/intro/language/python_language.html - Scipy Lectures: Lecture 1.2

## Versions

In [129]:
%reload_ext version_information
%version_information

Software versions
Python 3.5.1 64bit [GCC 4.2.1 (Apple Inc. build 5577)]
IPython 4.0.3
OS Darwin 15.0.0 x86_64 i386 64bit
Sun Mar 20 16:53:36 2016 EDT